File: python/test_harness.py
"""
A general-purpose testing and benchmarking harness for competitive programming problems.

Usage:
    from test_harness import TestHarness, Case

    TestHarness("Problem Name") \\
        .add_test("test name", lambda: solve(input), expected_output) \\
        .add_benchmark("large input", lambda: solve(large_input)) \\
        .add_scaling("N", sizes, generator_fn, solver_fn) \\
        .add_random_tests("Stress", 1000, case_generator_fn) \\
        .run()

Run any question:
    python question1_cow_sightings.py
"""

import time
import tracemalloc
from dataclasses import dataclass
from typing import Any, Callable


@dataclass
class Case:
    """A random test case generated at runtime."""
    description: str
    fn: Callable
    expected: Any


class TestHarness:

    WARMUP_RUNS = 3
    BENCHMARK_RUNS = 10
    SCALING_RUNS = 5
    MAX_FAILURES_SHOWN = 3

    def __init__(self, title: str):
        self._title = title
        self._steps: list[Callable] = []
        self._step_is_quick: list[bool] = []
        self._test_count = 0
        self._pass_count = 0
        self._step_index = 0

    # -- Adding tests & benchmarks ----------------------------

    def add_test(self, name_or_fn, fn=None, expected=None) -> "TestHarness":
        """Add a correctness test. Name is optional -- omit for auto-naming.

        Usage:
            .add_test("My test", lambda: solve(...), expected)
            .add_test(lambda: solve(...), expected)        # auto-named "Test N"
        """
        if callable(name_or_fn):
            # Called as add_test(fn, expected) -- no name
            name, fn, expected = None, name_or_fn, fn
        else:
            name = name_or_fn
        def step(name=name, fn=fn, expected=expected):
            self._test_count += 1
            if self._execute_test(name, fn, expected):
                self._pass_count += 1
        self._steps.append(step)
        self._step_is_quick.append(True)
        return self

    def add_benchmark(self, name_or_fn, fn=None) -> "TestHarness":
        """Add a benchmark. Name is optional -- omit for auto-naming.

        Usage:
            .add_benchmark("My bench", lambda: solve(...))
            .add_benchmark(lambda: solve(...))             # auto-named "Benchmark N"
        """
        if callable(name_or_fn):
            name, fn = None, name_or_fn
        else:
            name = name_or_fn
        self._steps.append(lambda name=name, fn=fn: self._execute_benchmark(name, fn))
        self._step_is_quick.append(False)
        return self

    def add_scaling(self, label: str, sizes: list[int],
                    generator, solver, reference=None) -> "TestHarness":
        """
        Add a scaling analysis that measures how solve time grows with input size.
        If reference is provided, shows side-by-side timing and memory comparison.

        Args:
            label:     description of the size parameter (e.g. "N (array size)")
            sizes:     list of input sizes to test, in ascending order
            generator: function(n) -> input of size n
            solver:    function(input) -> result
            reference: optional function(input) -> result for comparison
        """
        self._steps.append(
            lambda: self._execute_scaling(label, sizes, generator, solver, reference))
        self._step_is_quick.append(False)
        return self

    def add_random_tests(self, label: str, count: int,
                         case_generator: Callable[[], Case]) -> "TestHarness":
        """
        Add randomised stress tests comparing student solution vs reference.

        Args:
            label:          description of the stress test
            count:          number of random cases to generate
            case_generator: callable that returns a fresh Case each call
        """
        self._steps.append(
            lambda: self._execute_random_tests(label, count, case_generator))
        self._step_is_quick.append(True)
        return self

    # -- Running ----------------------------------------------

    def run(self) -> bool:
        """Run all steps. Returns True if every test passed."""
        return self._run_internal(quick_only=False)

    def run_quick(self) -> bool:
        """Run only correctness and random tests (skip benchmarks and scaling)."""
        return self._run_internal(quick_only=True)

    def _run_internal(self, quick_only: bool) -> bool:
        self._print_header()
        if quick_only:
            print("  (quick mode -- skipping benchmarks & scaling)\n")
        self._test_count = 0
        self._pass_count = 0
        self._step_index = 0

        for i, step in enumerate(self._steps):
            if quick_only and not self._step_is_quick[i]:
                continue
            self._step_index += 1
            step()
            print()

        self._print_summary(self._pass_count, self._test_count)
        return self._test_count > 0 and self._pass_count == self._test_count

    # -- Execution: correctness test --------------------------

    def _execute_test(self, name, fn, expected) -> bool:
        if name:
            print(f"  Test {self._step_index}: {name}")
        else:
            print(f"  Test {self._step_index}")

        for _ in range(self.WARMUP_RUNS):
            fn()

        start = time.perf_counter_ns()
        result = fn()
        elapsed_ns = time.perf_counter_ns() - start

        passed = result == expected

        if passed:
            print("    Status:   PASS")
        else:
            print("    Status:   FAIL")
            print(f"    Expected: {expected}")
            print(f"    Got:      {result}")
        print(f"    Time:     {elapsed_ns / 1_000_000:.3f} ms")

        return passed

    # -- Execution: benchmark ---------------------------------

    def _execute_benchmark(self, name, fn):
        if name:
            print(f"  Benchmark {self._step_index}: {name}")
        else:
            print(f"  Benchmark {self._step_index}")

        for _ in range(self.WARMUP_RUNS):
            fn()

        times = []
        for _ in range(self.BENCHMARK_RUNS):
            start = time.perf_counter_ns()
            fn()
            elapsed = time.perf_counter_ns() - start
            times.append(elapsed)

        avg_ms = (sum(times) / len(times)) / 1_000_000
        min_ms = min(times) / 1_000_000
        max_ms = max(times) / 1_000_000

        print(f"    Avg time: {avg_ms:.3f} ms  ({self.BENCHMARK_RUNS} runs)")
        print(f"    Min time: {min_ms:.3f} ms")
        print(f"    Max time: {max_ms:.3f} ms")

    # -- Execution: scaling analysis --------------------------

    def _execute_scaling(self, label: str, sizes: list[int],
                         generator, solver, reference=None):
        has_ref = reference is not None
        if has_ref:
            print(f"  Scaling {self._step_index}: {label} [vs reference]")
        else:
            print(f"  Scaling {self._step_index}: {label}")

        prev_ms = -1.0
        prev_size = -1

        for size in sizes:
            inp = generator(size)

            # Warmup solve
            for _ in range(self.WARMUP_RUNS):
                solver(inp)

            # Time solve
            total_ns = 0
            for _ in range(self.SCALING_RUNS):
                start = time.perf_counter_ns()
                solver(inp)
                total_ns += time.perf_counter_ns() - start
            solve_ms = (total_ns / self.SCALING_RUNS) / 1_000_000

            if not has_ref:
                if prev_ms <= 0:
                    print(f"    N = {size:>10,} | {solve_ms:>10.3f} ms")
                else:
                    time_ratio = solve_ms / prev_ms if prev_ms > 0 else 0
                    size_ratio = size / prev_size
                    print(f"    N = {size:>10,} | {solve_ms:>10.3f} ms"
                          f" | {time_ratio:>5.1f}x time for {size_ratio:>5.1f}x N")
            else:
                # Warmup + time reference
                for _ in range(self.WARMUP_RUNS):
                    reference(inp)
                ref_ns = 0
                for _ in range(self.SCALING_RUNS):
                    start = time.perf_counter_ns()
                    reference(inp)
                    ref_ns += time.perf_counter_ns() - start
                ref_ms = (ref_ns / self.SCALING_RUNS) / 1_000_000

                # Memory measurement via tracemalloc
                solve_mem = self._measure_memory(lambda: solver(inp))
                ref_mem = self._measure_memory(lambda: reference(inp))

                speedup = ref_ms / solve_ms if solve_ms > 0 else 0
                mem_str = f" | mem: {self._format_bytes(solve_mem)} vs {self._format_bytes(ref_mem)}"

                print(f"    N = {size:>10,} | solve: {solve_ms:>8.3f} ms"
                      f" | ref: {ref_ms:>8.3f} ms | {speedup:>5.1f}x{mem_str}")

            prev_ms = solve_ms
            prev_size = size

        print()
        if not has_ref:
            print("    Interpretation guide:")
            print("      ~1x time for kx N  =>  O(1)")
            print("      ~kx time for kx N  =>  O(N)")
            print("      ~k^2x for kx N     =>  O(N^2)")
        else:
            print("    speedup = ref_time / solve_time (higher is better)")
            print("    mem = peak memory allocated per call")

    # -- Execution: random stress tests -----------------------

    def _execute_random_tests(self, label: str, count: int,
                              case_generator: Callable[[], Case]):
        print(f"  Random Tests {self._step_index}: {label} ({count:,} cases)")

        passed = 0
        failures: list[str] = []

        for _ in range(count):
            tc = case_generator()
            result = tc.fn()
            if result == tc.expected:
                passed += 1
            elif len(failures) < self.MAX_FAILURES_SHOWN:
                failures.append(
                    f"    FAIL #{len(failures)+1}: {tc.description}\n"
                    f"      Expected: {tc.expected}\n"
                    f"      Got:      {result}")

        print(f"    Passed: {passed:,} / {count:,}")
        self._test_count += count
        self._pass_count += passed

        for f in failures:
            print(f)
        if count - passed > self.MAX_FAILURES_SHOWN:
            print(f"    ... and {count - passed - self.MAX_FAILURES_SHOWN:,} more failures")

    # -- Memory measurement helpers --------------------------

    @staticmethod
    def _measure_memory(fn) -> int:
        """Measure peak memory allocated by fn() using tracemalloc."""
        tracemalloc.start()
        fn()
        _, peak = tracemalloc.get_traced_memory()
        tracemalloc.stop()
        return peak

    @staticmethod
    def _format_bytes(b: int) -> str:
        if b < 1024:
            return f"{b} B"
        if b < 1024 * 1024:
            return f"{b / 1024:.1f} KB"
        return f"{b / (1024 * 1024):.1f} MB"

    # -- Formatting helpers -----------------------------------

    def _print_header(self):
        border = "=" * 55
        print(border)
        print(f"  Problem: {self._title}")
        print(border)
        print()

    def _print_summary(self, passed: int, total: int):
        border = "=" * 55
        print(border)
        if total == 0:
            print("  No correctness tests defined (benchmarks only).")
        else:
            print(f"  Summary: {passed:,} / {total:,} tests passed")
        print(border)

File: java/TestHarness.java
import java.util.*;
import java.util.function.*;
import java.lang.management.ManagementFactory;

/**
 * A general-purpose testing and benchmarking harness for competitive programming problems.
 *
 * Usage:
 *   TestHarness.&lt;ReturnType&gt;forProblem("Problem Name")
 *       .addTest("test name", () -&gt; solve(input), expectedOutput)
 *       .addBenchmark("large input", () -&gt; solve(largeInput))
 *       .addScaling("N", sizes, n -&gt; genInput(n), input -&gt; solve(input))
 *       .addRandomTests("Stress", 1000, () -&gt; new Case&lt;&gt;(...))
 *       .run();
 *
 * Compile & run any question:
 *   javac *.java
 *   java Question1_CowSightings
 */
public class TestHarness<T> {

    private static final int WARMUP_RUNS = 3;
    private static final int BENCHMARK_RUNS = 10;
    private static final int SCALING_RUNS = 5;
    private static final int MAX_FAILURES_SHOWN = 3;

    private final String problemTitle;
    private final List<Runnable> steps = new ArrayList<>();
    private final List<Boolean> stepIsQuick = new ArrayList<>();
    private int testCount = 0;
    private int passCount = 0;
    private int stepIndex = 0;

    /** A random test case generated at runtime. */
    public static class Case<T> {
        public final String description;
        public final Supplier<T> fn;
        public final T expected;

        public Case(String description, Supplier<T> fn, T expected) {
            this.description = description;
            this.fn = fn;
            this.expected = expected;
        }
    }

    // -- Factory ----------------------------------------------

    public static <T> TestHarness<T> forProblem(String title) {
        return new TestHarness<>(title);
    }

    private TestHarness(String title) {
        this.problemTitle = title;
    }

    // -- Adding tests & benchmarks ----------------------------

    /** Add a correctness test with an expected result. */
    public TestHarness<T> addTest(String name, Supplier<T> fn, T expected) {
        steps.add(() -> {
            testCount++;
            if (executeTest(name, fn, expected)) {
                passCount++;
            }
        });
        stepIsQuick.add(true);
        return this;
    }

    /** Add a correctness test (auto-named "Test N"). */
    public TestHarness<T> addTest(Supplier<T> fn, T expected) {
        return addTest(null, fn, expected);
    }

    /** Add a benchmark (no correctness check, just timing). */
    public TestHarness<T> addBenchmark(String name, Supplier<T> fn) {
        steps.add(() -> executeBenchmark(name, fn));
        stepIsQuick.add(false);
        return this;
    }

    /** Add a benchmark (auto-named "Benchmark N"). */
    public TestHarness<T> addBenchmark(Supplier<T> fn) {
        return addBenchmark(null, fn);
    }

    /**
     * Add a scaling analysis that measures how solve time grows with input size.
     *
     * @param label     description of the size parameter (e.g. "N (array size)")
     * @param sizes     array of input sizes to test, in ascending order
     * @param generator function that creates an input of the given size
     * @param solver    function that runs the solution on the generated input
     */
    public <I> TestHarness<T> addScaling(String label, int[] sizes,
            IntFunction<I> generator, Function<I, T> solver) {
        steps.add(() -> executeScaling(label, sizes, generator, solver, null));
        stepIsQuick.add(false);
        return this;
    }

    /**
     * Add a scaling analysis with a reference solution for side-by-side comparison.
     * Shows timing speedup and approximate memory usage at each input size.
     */
    public <I> TestHarness<T> addScaling(String label, int[] sizes,
            IntFunction<I> generator, Function<I, T> solver, Function<I, T> reference) {
        steps.add(() -> executeScaling(label, sizes, generator, solver, reference));
        stepIsQuick.add(false);
        return this;
    }

    /**
     * Add randomised stress tests that compare the student solution against
     * a known-correct reference.
     *
     * @param label         description of the stress test
     * @param count         number of random cases to generate
     * @param caseGenerator supplier that produces a fresh Case each call
     */
    public TestHarness<T> addRandomTests(String label, int count,
            Supplier<Case<T>> caseGenerator) {
        steps.add(() -> executeRandomTests(label, count, caseGenerator));
        stepIsQuick.add(true);
        return this;
    }

    // -- Running ----------------------------------------------

    /** Run all steps and return true if every test passed. */
    public boolean run() {
        return run(false);
    }

    /** Run only correctness and random tests (skip benchmarks and scaling). */
    public boolean runQuick() {
        return run(true);
    }

    private boolean run(boolean quickOnly) {
        printHeader();
        if (quickOnly) {
            System.out.println("  (quick mode -- skipping benchmarks & scaling)\n");
        }
        testCount = 0;
        passCount = 0;
        stepIndex = 0;

        for (int i = 0; i < steps.size(); i++) {
            if (quickOnly && !stepIsQuick.get(i)) {
                continue;
            }
            stepIndex++;
            steps.get(i).run();
            System.out.println();
        }

        printSummary(passCount, testCount);
        return testCount > 0 && passCount == testCount;
    }

    // -- Execution: correctness test -------------------------

    private boolean executeTest(String name, Supplier<T> fn, T expected) {
        if (name != null && !name.isEmpty()) {
            System.out.printf("  Test %d: %s%n", stepIndex, name);
        } else {
            System.out.printf("  Test %d%n", stepIndex);
        }

        for (int w = 0; w < WARMUP_RUNS; w++) {
            fn.get();
        }

        long startNs = System.nanoTime();
        T result = fn.get();
        long elapsedNs = System.nanoTime() - startNs;

        boolean passed = Objects.deepEquals(result, expected);

        if (passed) {
            System.out.printf("    Status:   PASS%n");
        } else {
            System.out.printf("    Status:   FAIL%n");
            System.out.printf("    Expected: %s%n", format(expected));
            System.out.printf("    Got:      %s%n", format(result));
        }
        System.out.printf("    Time:     %.3f ms%n", elapsedNs / 1_000_000.0);

        return passed;
    }

    // -- Execution: benchmark --------------------------------

    private void executeBenchmark(String name, Supplier<T> fn) {
        if (name != null && !name.isEmpty()) {
            System.out.printf("  Benchmark %d: %s%n", stepIndex, name);
        } else {
            System.out.printf("  Benchmark %d%n", stepIndex);
        }

        for (int w = 0; w < WARMUP_RUNS; w++) {
            fn.get();
        }

        long[] times = new long[BENCHMARK_RUNS];
        for (int r = 0; r < BENCHMARK_RUNS; r++) {
            long startNs = System.nanoTime();
            fn.get();
            times[r] = System.nanoTime() - startNs;
        }

        long min = Long.MAX_VALUE, max = Long.MIN_VALUE, sum = 0;
        for (long t : times) {
            min = Math.min(min, t);
            max = Math.max(max, t);
            sum += t;
        }
        double avg = sum / (double) BENCHMARK_RUNS;

        System.out.printf("    Avg time: %.3f ms  (%d runs)%n", avg / 1_000_000.0, BENCHMARK_RUNS);
        System.out.printf("    Min time: %.3f ms%n", min / 1_000_000.0);
        System.out.printf("    Max time: %.3f ms%n", max / 1_000_000.0);
    }

    // -- Execution: scaling analysis -------------------------

    private <I> void executeScaling(String label, int[] sizes,
            IntFunction<I> generator, Function<I, T> solver, Function<I, T> reference) {
        boolean hasRef = reference != null;
        if (hasRef) {
            System.out.printf("  Scaling %d: %s [vs reference]%n", stepIndex, label);
        } else {
            System.out.printf("  Scaling %d: %s%n", stepIndex, label);
        }

        double prevMs = -1;
        int prevSize = -1;

        for (int size : sizes) {
            I input = generator.apply(size);

            // Warmup solve
            for (int w = 0; w < WARMUP_RUNS; w++) {
                solver.apply(input);
            }

            // Time solve
            long solveSum = 0;
            for (int r = 0; r < SCALING_RUNS; r++) {
                long startNs = System.nanoTime();
                solver.apply(input);
                solveSum += System.nanoTime() - startNs;
            }
            double solveMs = (solveSum / (double) SCALING_RUNS) / 1_000_000.0;

            if (!hasRef) {
                if (prevMs <= 0) {
                    System.out.printf("    N = %,10d | %10.3f ms%n", size, solveMs);
                } else {
                    double timeRatio = solveMs / prevMs;
                    double sizeRatio = (double) size / prevSize;
                    System.out.printf("    N = %,10d | %10.3f ms | %5.1fx time for %5.1fx N%n",
                        size, solveMs, timeRatio, sizeRatio);
                }
            } else {
                // Warmup + time reference
                for (int w = 0; w < WARMUP_RUNS; w++) {
                    reference.apply(input);
                }
                long refSum = 0;
                for (int r = 0; r < SCALING_RUNS; r++) {
                    long startNs = System.nanoTime();
                    reference.apply(input);
                    refSum += System.nanoTime() - startNs;
                }
                double refMs = (refSum / (double) SCALING_RUNS) / 1_000_000.0;

                // Memory: single-run measurement for each
                long solveMem = measureAllocatedBytes(() -> solver.apply(input));
                System.gc();
                long refMem = measureAllocatedBytes(() -> reference.apply(input));

                double speedup = (solveMs > 0) ? refMs / solveMs : 0;

                String memStr = "";
                if (solveMem >= 0 && refMem >= 0) {
                    memStr = String.format(" | mem: %s vs %s",
                        formatBytes(solveMem), formatBytes(refMem));
                }

                System.out.printf("    N = %,10d | solve: %8.3f ms | ref: %8.3f ms | %5.1fx%s%n",
                    size, solveMs, refMs, speedup, memStr);
            }

            prevMs = solveMs;
            prevSize = size;
        }

        System.out.println();
        if (!hasRef) {
            System.out.println("    Interpretation guide:");
            System.out.println("      ~1x time for kx N  =>  O(1)");
            System.out.println("      ~kx time for kx N  =>  O(N)");
            System.out.println("      ~k^2x for kx N     =>  O(N^2)");
        } else {
            System.out.println("    speedup = ref_time / solve_time (higher is better)");
            System.out.println("    mem = approx. heap bytes allocated per call");
        }
    }

    // -- Execution: random stress tests ----------------------

    private void executeRandomTests(String label, int count,
            Supplier<Case<T>> caseGenerator) {
        System.out.printf("  Random Tests %d: %s (%,d cases)%n", stepIndex, label, count);

        int passed = 0;
        List<String> failures = new ArrayList<>();

        for (int i = 0; i < count; i++) {
            Case<T> tc = caseGenerator.get();
            T result = tc.fn.get();
            if (Objects.deepEquals(result, tc.expected)) {
                passed++;
            } else if (failures.size() < MAX_FAILURES_SHOWN) {
                failures.add(String.format(
                    "    FAIL #%d: %s%n      Expected: %s%n      Got:      %s",
                    failures.size() + 1, tc.description,
                    format(tc.expected), format(result)));
            }
        }

        System.out.printf("    Passed: %,d / %,d%n", passed, count);
        testCount += count;
        passCount += passed;

        for (String f : failures) {
            System.out.println(f);
        }
        if (count - passed > MAX_FAILURES_SHOWN) {
            System.out.printf("    ... and %,d more failures%n",
                count - passed - MAX_FAILURES_SHOWN);
        }
    }

    // -- Memory measurement helpers --------------------------

    /** Measure heap bytes allocated during a single function call via ThreadMXBean. */
    private static long measureAllocatedBytes(Runnable fn) {
        long before = getThreadAllocatedBytes();
        fn.run();
        long after = getThreadAllocatedBytes();
        return (before >= 0 && after >= 0) ? after - before : -1;
    }

    private static long getThreadAllocatedBytes() {
        try {
            com.sun.management.ThreadMXBean tmx =
                (com.sun.management.ThreadMXBean) ManagementFactory.getThreadMXBean();
            return tmx.getThreadAllocatedBytes(Thread.currentThread().getId());
        } catch (Exception e) {
            return -1;
        }
    }

    private static String formatBytes(long bytes) {
        if (bytes < 0) {
            return "N/A";
        }
        if (bytes < 1024) {
            return bytes + " B";
        }
        if (bytes < 1024 * 1024) {
            return String.format("%.1f KB", bytes / 1024.0);
        }
        return String.format("%.1f MB", bytes / (1024.0 * 1024));
    }

    // -- Formatting helpers -----------------------------------

    private static String repeatStr(String s, int n) {
        StringBuilder sb = new StringBuilder(s.length() * n);
        for (int i = 0; i < n; i++) {
            sb.append(s);
        }
        return sb.toString();
    }

    private void printHeader() {
        String border = repeatStr("=", 55);
        System.out.println(border);
        System.out.printf("  Problem: %s%n", problemTitle);
        System.out.println(border);
        System.out.println();
    }

    private void printSummary(int passed, int total) {
        String border = repeatStr("=", 55);
        System.out.println(border);
        if (total == 0) {
            System.out.println("  No correctness tests defined (benchmarks only).");
        } else {
            System.out.printf("  Summary: %,d / %,d tests passed%n", passed, total);
        }
        System.out.println(border);
    }

    /** Pretty-print any result type, including primitive arrays. */
    private static String format(Object obj) {
        if (obj == null) {
            return "null";
        }
        if (obj instanceof int[]) {
            return Arrays.toString((int[]) obj);
        }
        if (obj instanceof long[]) {
            return Arrays.toString((long[]) obj);
        }
        if (obj instanceof double[]) {
            return Arrays.toString((double[]) obj);
        }
        if (obj instanceof boolean[]) {
            return Arrays.toString((boolean[]) obj);
        }
        if (obj instanceof char[]) {
            return Arrays.toString((char[]) obj);
        }
        if (obj instanceof Object[]) {
            return Arrays.deepToString((Object[]) obj);
        }
        return obj.toString();
    }
}
